x/config/decode.go

```diff
@@ -0,0 +1,216 @@
+	if err := decode(&items); err != nil {
+		return fmt.Errorf("failed to decode inbound items: %v", err)
+	}
+
+	for k, v := range items {
+		if v.Type == "" {
+			v.Type = k
+		}
+		*is = append(*is, v)
+	}
+	return nil
+}
+
+type inbound struct {
+	Type       string
+	Disabled   bool
+	Attributes attributeMap
+}
+
+func (i *inbound) Decode(decode decode.Into) error {
+	if err := decode(&i.Attributes); err != nil {
+		return fmt.Errorf("failed to decode inbound: %v", err)
+	}
+
+	var err error
+	i.Type, err = i.Attributes.PopString("type")
+	if err != nil {
+		return fmt.Errorf(`failed to read attribute "type" of inbound: %v`, err)
+	}
+	i.Disabled, err = i.Attributes.PopBool("disabled")
+	if err != nil {
+		return fmt.Errorf(`failed to read attribute "disabled" of inbound: %v`, err)
+	}
+
+	return nil
+}
+
+type clientConfigs map[string]outbounds
+
+func (cc *clientConfigs) Decode(decode decode.Into) error {
+	var items map[string]outbounds
+	if err := decode(&items); err != nil {
+		return fmt.Errorf("failed to decode outbound items: %v", err)
+	}
+
+	for k, v := range items {
+		if v.Service == "" {
+			v.Service = k
+		}
+		items[k] = v
+	}
+	*cc = items
+	return nil
+}
+
+type outbounds struct {
+	Service string
+
+	// Either (Unary and/or Oneway) will be set or Implicit will be set. For
+	// the latter case, we need to only use those configurations that that
+	// transport supports.
+	Unary    *outbound
+	Oneway   *outbound
+	Implicit *outbound
+}
+
+func (o *outbounds) Decode(decode decode.Into) error {
+	var attrs attributeMap
+	if err := decode(&attrs); err != nil {
+		return fmt.Errorf("failed to decode outbound configuration: %v", err)
+	}
+
+	var err error
+	o.Service, err = attrs.PopString("service")
+	if err != nil {
+		return fmt.Errorf("failed to read service name for outbound: %v", err)
+	}
+
+	hasUnary, err := attrs.Pop("unary", &o.Unary)
+	if err != nil {
+		return fmt.Errorf("failed to unary outbound configuration: %v", err)
+	}
+
+	hasOneway, err := attrs.Pop("oneway", &o.Oneway)
+	if err != nil {
+		return fmt.Errorf("failed to oneway outbound configuration: %v", err)
+	}
+
+	if hasUnary || hasOneway {
+		// No more attributes should be remaining
+		var empty struct{}
+		if err := attrs.Decode(&empty); err != nil {
+			return fmt.Errorf(
+				"too many attributes in explicit outbound configuration: %v", err)
+		}
+		return nil
+	}
+
+	if err := attrs.Decode(&o.Implicit); err != nil {
+		return fmt.Errorf("failed to decode implicit outbound configuration: %v", err)
+	}
+	return nil
+}
+
+type outbound struct {
+	Type       string
+	Preset     string
+	Attributes attributeMap
+}
+
+func (o *outbound) Decode(decode decode.Into) error {
+	var cfg map[string]attributeMap
+	if err := decode(&cfg); err != nil {
+		return fmt.Errorf("failed to decode outbound: %v", err)
+	}
+
+	switch len(cfg) {
+	case 0:
+		return errors.New("failed to decode outbound: an outbound type is required")
+	case 1:
+		// Move along
```

Comment:
No. That case is handled outside the switch statement. It's a largish-blob of logic so I didn't want it to be nested.